### PR TITLE
Making the render engine Job/Task system configurable and reusable

### PR DESCRIPTION
**Pull request by samcake**
In order to reuse the Job/Task classes system to create different kind of engines we had to clean up the interface and make it configurable and separated from the rendering specifics.

We separated the task code in its own namespace "task" still in the _render_ libraries and partitioned the big task.h file into 3 files Varying.h, Config.h/.cpp and Task.h.
The _run_ method of the job has now that Context type that can be customized which allow for its specialization 

The render/Engine file now contains the instantiation of the task with the RenderContext.
A simple macro allow for aliasing all the required types of "task" into "render"

## TEST PLAN
Nothing changed in behavior or performances with this pr.
So a simple check for identical performances and behavior of rendering compared to Master should be enough.
Test enabling /disabling shadows and Ambient Occlusion (in the developer/graphics menu). Make sure that the states are recovered correctly from sessions to sessions.

### CODE REVIEW

File: libraries/render-utils/src/AmbientOcclusionEffect.h [hunk 1]
```diff
@@ -71,7 +71,7 @@ class AmbientOcclusionEffectConfig : public render::GPUJobConfig::Persistent {
     Q_PROPERTY(int blurRadius MEMBER blurRadius WRITE setBlurRadius)
 
 public:
-    AmbientOcclusionEffectConfig() : render::GPUJobConfig::Persistent("Ambient Occlusion", false) {}
+    AmbientOcclusionEffectConfig() : render::GPUJobConfig::Persistent(QStringList() << "Render" << "Engine" << "Ambient Occlusion", false) {}
```
Comment (samcake) [inline libraries/render-utils/src/AmbientOcclusionEffect.h]: I removed the Render/Engine from the default PersistentConfiguration class and so have to do it there

File: libraries/render-utils/src/DeferredLightingEffect.h [hunk 2]
```diff
@@ -20,7 +20,6 @@
 #include "model/Light.h"
 #include "model/Geometry.h"
 
-#include "render/Context.h"
```
Comment (samcake) [inline libraries/render-utils/src/DeferredLightingEffect.h]: No more Context file the same class is now in task/Task.h

File: libraries/render-utils/src/RenderShadowTask.h [hunk 3]
```diff
@@ -35,7 +35,7 @@ class RenderShadowTaskConfig : public render::Task::Config::Persistent {
     Q_OBJECT
     Q_PROPERTY(bool enabled MEMBER enabled NOTIFY dirty)
 public:
-    RenderShadowTaskConfig() : render::Task::Config::Persistent("Shadows", false) {}
+    RenderShadowTaskConfig() : render::Task::Config::Persistent(QStringList() << "Render" << "Engine" << "Shadows", false) {}
```
Comment (samcake) [inline libraries/render-utils/src/RenderShadowTask.h]: Same as above, the PersistentConfiguration class constructor doesn't add Render/Engine to the SettingHandle path, so we do it here

File: libraries/render/src/render/Context.h [hunk 4]
```diff
@@ -1,32 +0,0 @@
...
-    RenderArgs* args;
-    std::shared_ptr<JobConfig> jobConfig{ nullptr };
-    ScenePointer _scene;
-};
-using RenderContextPointer = std::shared_ptr<RenderContext>;
-
-}
-
-#endif // hifi_render_Context_h
-
```
Comment (samcake) [inline libraries/render/src/render/Context.h]: All this classes are now in task/Task.h and then in the Engine.h file right next to the instanciation of the task system for rendering

File: libraries/render/src/render/Engine.h [hunk 5]
```diff
@@ -14,10 +14,67 @@
...
-#include "Task.h"
+#include "Scene.h"
+#include "../task/Task.h"
+#include "gpu/Batch.h"
+
 namespace render {
 
+
+
+    class RenderContext : public task::JobContext {
```
Comment (samcake) [inline libraries/render/src/render/Engine.h]: RenderContext must derive from the base class task::JobContent

File: libraries/render/src/render/Engine.h [hunk 6]
```diff
@@ -14,10 +14,67 @@
...
+    public:
+        virtual ~RenderContext() {}
+
+        RenderArgs* args;
+        ScenePointer _scene;
+    };
+    using RenderContextPointer = std::shared_ptr<RenderContext>;
+
+    Task_DeclareTypeAliases(RenderContext)
```
Comment (samcake) [inline libraries/render/src/render/Engine.h]: This Macro is where all the required classes and types from task/ are instantiated with the RenderContext and aliased in the current namespace aka "render"

File: libraries/render/src/render/Engine.h [hunk 7]
```diff
@@ -14,10 +14,67 @@
...
+
+
+        GPUTaskConfig() = default;
+        GPUTaskConfig(bool enabled) : TaskConfig(enabled) {}
+
+        // Running Time measurement on GPU and for Batch execution
+        void setGPUBatchRunTime(double msGpuTime, double msBatchTime) { _msGPURunTime = msGpuTime; _msBatchRunTime = msBatchTime; }
+        double getGPURunTime() const { return _msGPURunTime; }
+        double getBatchRunTime() const { return _msBatchRunTime; }
+    };
```
Comment (samcake) [inline libraries/render/src/render/Engine.h]: The GPUJobConfig and GPUTaskConfig are specific to render so they moved here

File: libraries/render/src/task/Config.cpp [hunk 8]
```diff
@@ -1,19 +1,28 @@
...
 
-using namespace render;
+void JobConfig::setPresetList(const QJsonObject& object) {
+    for (auto it = object.begin(); it != object.end(); it++) {
+        JobConfig* child = findChild<JobConfig*>(it.key(), Qt::FindDirectChildrenOnly);
+        if (child) {
+            child->setPresetList(it.value().toObject());
+        }
+    }
+}
```
Comment (samcake) [inline libraries/render/src/task/Config.cpp]: Moved this function 's body in the cpp

File: libraries/render/src/task/Task.h [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+    std::shared_ptr<Config> getConfiguration() {
+        return std::static_pointer_cast<Config>(_concept->getConfiguration());
+    }
+
+protected:
+};
+}
+
+
+#define Task_DeclareTypeAliases(ContextName) \
```
Comment (samcake) [inline libraries/render/src/task/Task.h]: The Macro allowing to instantiate the task namespace types with the specified context and also aliased in the namespace where it s called.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 33d770c24e27fb3fb4eafe2627e134bf08095482 available at: https://gist.github.com/cad48220324308733d24359db1bb6dfa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-33d770c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-33d770c.dmg

**Comment by Nex-Pro** [comment]
Testing this

**Comment by Nex-Pro** [comment]
### I think this is a fail.
There is a significant change in the rendering of colours. I compared this to 6353

In the image below I have indicated the edges of a section of an image made in 6353 overlaid on one from this PR. There is a marked difference in tone.  Images were made less than two minutes apart.

shading change

**Comment by Nex-Pro** [comment]
Ambient occlusion and shadow states persist correctly across restarts. Changing from AO and shadow on to off occurs quickly.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c36f8da4c00afbdf0c3af5ca4365cf021a168758 available at: https://gist.github.com/cad48220324308733d24359db1bb6dfa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-c36f8da.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-c36f8da.dmg

**Comment by samcake** [comment]
@Nex-Pro Hi Piper,
Thanks for the review
Concerning the tone change, i don;t really repro on my side.
I m testing with almost the same content (the tree) but in blue (my domain).
I don't see any difference between current master 6340 and this pr:
capture-master6340
capture-pr10265

**Comment by samcake** [comment]
I m will test now with the actual master 6354

**Comment by samcake** [comment]
And in the latest master 6354:
capture-master6354
It looks all identical to me.

I think the lighting is changing constantly in Welcome so it could be why.
also i realized that the trees are moving ?
But your snapshot the tree is stable relative to the avatar, so is it a non moving tree ?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2d3ebace06090cce6fa6b859587e24a6356a8f76 available at: https://gist.github.com/cad48220324308733d24359db1bb6dfa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-2d3ebac.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-2d3ebac.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b1d58061f11636551e80f9dbb2c5dadaeb8849f2 available at: https://gist.github.com/cad48220324308733d24359db1bb6dfa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-b1d5806.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-b1d5806.dmg

**Comment by Nex-Pro** [comment]
@samcake  I tried to get the exact same position and as close in time as possible to the same frame of the tree animation as possible to make sure that the tree animation was not causing the issue.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4514da2a09f918606792b476358ff6e96e3f492c available at: https://gist.github.com/cad48220324308733d24359db1bb6dfa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-4514da2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-4514da2.dmg

**Comment by samcake** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4514da2a09f918606792b476358ff6e96e3f492c available at: https://gist.github.com/cad48220324308733d24359db1bb6dfa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-4514da2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-4514da2.dmg

**Comment by samcake** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4514da2a09f918606792b476358ff6e96e3f492c available at: https://gist.github.com/cad48220324308733d24359db1bb6dfa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-4514da2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-4514da2.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: df54ad233977f58acc1cd6ae61d89b96f1671f5d available at: https://gist.github.com/cad48220324308733d24359db1bb6dfa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-df54ad2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-df54ad2.dmg

**Comment by samcake** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: df54ad233977f58acc1cd6ae61d89b96f1671f5d available at: https://gist.github.com/cad48220324308733d24359db1bb6dfa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-df54ad2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10265-df54ad2.dmg

**Comment by samcake** [comment]
Tested on various platform (nvidia i7, macos, amd i5)
It s a pass

**Pull request merged**